### ISSUE TITLE
Bad state: Can't finalize a finalized MultipartFile

### ISSUE THREAD

**Issue by aweiand** [rodion-m/dio_smart_retry#15]
Hi,

I'm using the latest packages:
```
dio: ^4.0.6
dio_smart_retry: ^1.3.2
```
When I post a file to the server and, it accepts only after first attemp, the error occours... here a simplified code:

```
    Dio dio = Dio(BaseOptions(
      baseUrl: ("https://$domain$base"),
    );

    dio.interceptors.add(
        RetryInterceptor(
          dio: dio,
          logPrint: print, // specify log function (optional)
          retries: 3, // retry count (optional)
          retryDelays: const [
            // set delays between retries (optional)
            Duration(seconds: 1), // wait 1 sec before first retry
            Duration(seconds: 2), // wait 2 sec before second retry
            Duration(seconds: 3), // wait 3 sec before third retry
          ],
        ),
      );

    final bytes = await cameraFile!.readAsBytes();
    final MultipartFile file = MultipartFile.fromBytes(bytes, filename: "teste");
    MapEntry<String, MultipartFile> a = MapEntry('teste', file);

    var options = FormData.fromMap({
      "bla": "bla",
    });

    options.files.add(a);

    var data = await dio.post("/post_test", data: options);
```

**Comment by vimaxwell** [comment]
please test my PR https://github.com/rodion-m/dio_smart_retry/pull/16

**Comment by aweiand** [comment]
Hi @vimaxwell ,

After test in a test server I own, the issue persists...

**Comment by wolru** [comment]
@aweiand please test this PR
https://github.com/rodion-m/dio_smart_retry/pull/17

I'm not using this library, so I edited with web editor. 
If there is problem, let me know

**Comment by aweiand** [comment]
@wolru , unfortunately the PR #17 don't fix the issue....

Although, the PR occasionally the error occurs on the first, second... attempt, not only on the last

Here is the stacktrace

```
#0      MultipartFile.finalize (package:dio/src/multipart_file.dart:133:7)
#1      FormData.finalize.<anonymous closure> (package:dio/src/form_data.dart:164:43)
#2      Future.forEach.<anonymous closure> (dart:async/future.dart:646:26)
#3      Future.doWhile.<anonymous closure> (dart:async/future.dart:703:26)
#4      _RootZone.runUnaryGuarded (dart:async/zone.dart:1586:10)
#5      _RootZone.bindUnaryCallbackGuarded.<anonymous closure> (dart:async/zone.dart:1625:26)
#6      Future.doWhile (dart:async/future.dart:719:18)
#7      Future.forEach (dart:async/future.dart:644:12)
#8      FormData.finalize (package:dio/src/form_data.dart:161:12)
#9      DioMixin._transformData (package:dio/src/dio_mixin.dart:730:23)
#10     DioMixin._dispatchRequest (package:dio/src/dio_mixin.dart:656:26)
#11     DioMixin.fetch.<anonymous closure> (package:dio/src/dio_mixin.dart:605:7)
#12     DioMixin.fetch._requestInterceptorWrapper.<anonymous closure>.<anonymous closure>.<anonymous closure> (package:dio/src/dio_mixin.dart:517:28)
#13     DioMixin.checkIfNeedEnqueue (package:dio/src/dio_mixin.dart:789:22)
#14     DioMixin.fetch._requestInterceptorWrapper.<anonymous closure>.<anonymous closure> (package:dio/src/dio_mixin.dart:515:22)
#15     new Future.<anonymous closure> (dart:async/future.dart:253:37)
#16     Timer._createTimer.<anonymous closure> (dart:async-patch/timer_patch.dart:18:15)
#17     _Timer._runTimers (dart:isolate-patch/timer_impl.dart:398:19)
#18     _Timer._handleMessage (dart:isolate-patch/timer_impl.dart:429:5)
#19     _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:192:12)
```

And here, is the screenshot of the error with the original plugin's code:
![image](https://user-images.githubusercontent.com/1443817/204647847-1b6e417b-8521-4977-8ae5-77e0f78e9fe0.png)

**Comment by aweiand** [comment]
Ow, this is the PHP code (in Laravel) that I'm using to test (inside a controller and api route):

```
public function post_test(Request $request)
    {
        try {
            $file = $request->file('teste');
            $test = session('post_test');

            if($test == null || session('post_test') == 1){
                session('post_test') == null ? $request->session()->put('post_test', 1) : 
                                                $request->session()->put('post_test', 2);

                Log::alert('LOG 525', [
                    'files' => $file,
                    'request' => $request->all()
                ]);

                return response()->json([
                    'msg' => 'error',
                    'sess' => session('post_test')
                ], 525);
            } else {
                $request->session()->put('post_test', null);

                Log::alert('LOG 200', [
                    'files' => $file,
                    'request' => $request->all()
                ]);

                return response()->json([
                    'msg' => 'uhuu',
                    'sess' => session('post_test')
                ], 200);
            }
        } catch (\Exception $e) {
            Log::critical($e->getMessage(), ['file' => $e->getFile(), 'line' => $e->getLine(), 'request' => $request->all()]);
        }
    }
```

**Comment by rodion-m** [comment]
Hey @aweiand ! I've added the fix for this issue. Now to enable retrying you should add multipart data like this (using `MultipartFileRecreatable`):
```dart
    final formData =
        FormData.fromMap({'file': MultipartFileRecreatable.fromFileSync('README.md')});
      await dio.post<dynamic>(
        'https://multipart.free.beeceptor.com/post500',
        data: formData,
      );
```
See the full example in the test: https://github.com/rodion-m/dio_smart_retry/blob/63a3bddae8b5a0581c35c4ae5e973996561d9100/test/multipart_retry_tests.dart#L32\
Could you please check your case again using src from the current repo?

**Comment by rodion-m** [comment]
It's the best solution I've found for today, but it's not very good, because it's not supported `MultipartFile` class, since this class doesn't have a `filepath` field or even a public stream. I hope dio developers will add any solution for this issue that will allow rewinding a multipart data stream.

**Comment by rodion-m** [comment]
While dio doesn't support `MultipartFile` recreation I want to consult with the community - **should `RetryInterceptor` throw an exception before sending a request?** I think this is the right behavior, but it'll break backward compatibility. At the moment I'll add just a warning to the log about the issue. If somebody has thoughts about it - feel free to share it in this thread.

**Comment by aweiand** [comment]
@rodion-m don't work here.... it was sending the same error, I'm using this code:

´´´
  Future<dynamic> postTest() async {
    final MultipartFileRecreatable file = MultipartFileRecreatable.fromFileSync(cameraFile!.path, filename: "teste");
    MapEntry<String, MultipartFileRecreatable> a = MapEntry('teste', file);

    var options = FormData.fromMap({
      "bla": "bla",
    });

    options.files.add(a);

    var data = await todoKmHttp.post("/post_test", data: options);
    return data;
  }
´´´

In my opinion you don't has to throw an exception generically, maybe only on some listed error code?

**Comment by aweiand** [comment]
@rodion-m, in a short test, it was suceffull! I will design a full test in the sequence. Thank's!!

In my opinion you don't has to throw an exception generically, maybe only on some listed error code?

**Comment by rodion-m** [comment]
> in a short test, it was suceffull! I will design a full test in the sequence. Thank's!!

Ok, thank you for testing!

> In my opinion you don't has to throw an exception generically, maybe only on some listed error code?
If formdata has a type `MultipartFile` then the exception will be thrown anyway on retry (because a retry is not possible for `MultipartFile`).

**Comment by helloalbin** [comment]
Is It possible to add support for  fromString method as well? Currently if we have both data as well as Files in the same request, retry is not working. My current code is 

```
MultipartFile jsonData = MultipartFile.fromString(
          jsonEncode(data),
          contentType: http_parser.MediaType.parse('application/json'),
        );
```

So can we have an option like

```
MultipartFile jsonData = MultipartFileRecreatable.fromString(
          jsonEncode(data),
          contentType: http_parser.MediaType.parse('application/json'),
        );
```

**Comment by rodion-m** [comment]
Hey @helloalbin ! Certainly, feel free to add it as a PR :)

**Comment by rodion-m** [comment]
fixed with #42.
